src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import io.circe.{Codec, Decoder, Encoder}
+import io.circe.generic.semiauto.deriveCodec
+import io.circe.syntax.EncoderOps
+import org.constellation.ConstellationExecutionContext.createSemaphore
+import org.constellation.domain.healthcheck.HealthCheckConsensusManager.{
+  FetchProposalError,
+  FetchingFailed,
+  HistoricalRound,
+  SendProposalError,
+  SendingFailed
+}
+import org.constellation.domain.p2p.PeerHealthCheck.{
+  PeerAvailable,
+  PeerHealthCheckStatus,
+  PeerUnresponsive,
+  UnknownPeer
+}
+import org.constellation.infrastructure.p2p.{ClientInterpreter, PeerResponse}
+import org.constellation.p2p.PeerData
+import org.constellation.schema.{Id, NodeState}
+import org.constellation.schema.consensus.RoundId
+
+import scala.concurrent.duration.SECONDS
+import scala.util.{Random, Try}
+
+class HealthCheckConsensus[F[_]](
+  checkedId: Id,
+  ownId: Id,
+  roundId: RoundId,
+  initialRoundPeers: Map[Id, PeerData],
+  healthCheckStatus: PeerHealthCheckStatus,
+  startedAtSecond: Long,
+  val isKeepUpRound: Boolean
+)(
+  apiClient: ClientInterpreter[F],
+  unboundedBlocker: Blocker
+)(implicit F: Concurrent[F], CS: ContextShift[F], C: Clock[F]) {
+  import HealthCheckConsensus._
+
+  implicit val roundIdOrdering: Ordering[RoundId] = Ordering.by[RoundId, String](_.id)
+  implicit val roundIdOrder: Order[RoundId] = Order.fromOrdering
+  implicit val peerDataOrdering: Ordering[PeerData] = Ordering.by[PeerData, String](_.peerMetadata.id.hex)
+  implicit val peerDataOrder: Order[PeerData] = Order.fromOrdering
+
+  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  val fetchProposalAfter = 120L
+  val removeUnresponsivePeersWithParallelRoundAfter = 60L
+
+  private[healthcheck] val roundIds: Ref[F, NonEmptySet[RoundId]] = Ref.unsafe(NonEmptySet.one(roundId))
+  private[healthcheck] val roundPeers: Ref[F, Map[Id, RoundData]] = {
+    Ref.unsafe(
+      initialRoundPeers
+        .mapValues(pd => RoundData(pd.asRight[ProxyPeers]))
+    )
+  }
+  private[healthcheck] val leftDuringConsensus: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val removedUnresponsivePeersWithParallelRound: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val ownConsensusHealthStatus: Ref[F, ConsensusHealthStatus] = Ref.unsafe(
+    ConsensusHealthStatus(
+      checkedId,
+      ownId,
+      roundId,
+      healthCheckStatus,
+      initialRoundPeers.mapValues(_.peerMetadata.nodeState)
+    )
+  )
+  private[healthcheck] val parallelRounds: Ref[F, Map[Id, Set[RoundId]]] = Ref.unsafe(Map.empty)
+
+  private val sendingProposalLock = createSemaphore()
+  private val fetchingProposalLock = createSemaphore()
+
+  def getTimeInSeconds(): F[Long] = C.monotonic(SECONDS)
+
+  def start(): F[Fiber[F, Unit]] = F.start(sendProposal())
+
+  def getOwnPerceivedHealthStatus(): F[ConsensusHealthStatus] = ownConsensusHealthStatus.get
```

Comment:
If we have public accessors for these fields, maybe we shouldn't keep them in `[healthcheck]` protected scope at all, but rather just make them public?

Reply:
if we have public accessors, the underlying variable should be hidden imo

Reply:
I'm not sure yet, maybe we will make these accessors defs accessible only in [healthcheck] scope. These methods are just kind of meant for unified easy acces. For sure I wanted Ref's to be in [healtcheck] scope to be able to modify them in tests if not that they would probably be private.

Reply:
@tbekas regarding the meeting, I agree and I think it's needed. I also need you guys to understand it in order to help me find some shortcomings. We will have it tomorrow during the knowledge sharing session.